vllm/model_executor/guided_decoding/guidance_logits_processors.py

```diff
@@ -0,0 +1,189 @@
+        self.ll_tokenizer = self.metadata[
+            f"ll_tokenizer_{self.tokenizer_name}"]
+
+        self.ll_interpreter = llguidance.LLInterpreter(
+            self.ll_tokenizer,
+            json.dumps(self.serialized_grammar),
+            enable_backtrack=False,
+            log_level=int(os.environ.get("LLGUIDANCE_LOG_LEVEL", "1")),
+        )
+
+        self.initialized = True
+
+    def __call__(
+        self,
+        prompt_tokens_ids: List[int],
+        past_tokens_ids: list[int],
+        logits: torch.Tensor,
+    ) -> torch.Tensor:
+        # we initialize the guidance model here
+        # to avoid pickling ll_tokenizer and ll_interpreter
+        self._initialize()
+
+        if self.is_stopped:
+            return logits
+
+        try:
+            if len(past_tokens_ids) == 0:
+                self.ll_interpreter.process_prompt(prompt_tokens_ids)
+
+            if self.new_sampling and len(past_tokens_ids) > 0:
+                backtrack, ff_tokens = self.ll_interpreter.post_process(
+                    past_tokens_ids[-1])
+                if len(ff_tokens) > 0 and backtrack == 0:
+                    # first token is last generated token
+                    ff_tokens = ff_tokens[1:]
+                self.pending_ff_tokens.extend(ff_tokens)
+                self.new_sampling = False
+
+            if len(self.pending_ff_tokens) > 0:
+                # if we have pending fast-forward tokens,
+                # just return them immediately
+                ff_token = self.pending_ff_tokens.pop(0)
+                masked_logits = torch.zeros_like(logits,
+                                                 dtype=logits.dtype,
+                                                 device=logits.device)
+                masked_logits[ff_token] = 200.0
+                return masked_logits
+
+            mask, resp = self.ll_interpreter.mid_process()
+            r = LLInterpreterResponse.model_validate_json(resp)
+
+            if r.stop:
+                mask = torch.zeros_like(logits,
+                                        dtype=logits.dtype,
+                                        device=logits.device)
+                if self.guidance_tokenizer.eos_token_id is not None:
+                    mask[self.guidance_tokenizer.eos_token_id] = 200.0
+                self.is_stopped = True
+            elif mask is None:
+                # NOTE: mask should not be None unless r.stop is True
+                # However, we are handling this case just in case
+                # llguidance allows free-style generation
+                mask = torch.zeros_like(logits,
+                                        dtype=logits.dtype,
+                                        device=logits.device)
+            else:
+                mask = np.frombuffer(mask, dtype=np.uint8)
+                mask = torch.tensor(mask,
+                                    dtype=logits.dtype,
+                                    device=logits.device)
```

Comment:
Can the allocated mask tensor be reused between calls?

Reply:
Unfortunately, the mask is generated from the rust-based llguidance library so I cannot pass the tensor to the lib.
And I don't know if we have a good way to update existing tensor with a new numpy array? Any suggestion?